### PR TITLE
Add tools to update and generate tables and tests.

### PR DESCRIPTION
**Pull request by lrhn**
Include low-level benchmark in tool/.

### CODE REVIEW

File: tool/benchmark.dart [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import "package:characters/src/grapheme_clusters/breaks.dart";
+import "package:characters/src/grapheme_clusters/constants.dart";
+
+import "../test/src/text_samples.dart";
+import "../test/src/unicode_grapheme_tests.dart";
+import "../test/src/various_tests.dart";
+
+// Low-level benchmark of the grpaheme cluster step functions.
```
Comment (natebosch) [inline tool/benchmark.dart]:
```suggestion
// Low-level benchmark of the grapheme cluster step functions.
```

File: tool/bin/gentable.dart [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+//
+// The approach taken is to split the large table into chunks,
+// smaller chunks for data in the BMP (U+0000 ... U+FFFF)
+// which have higher variation than later data,
+// and larger chunks for non-BMP ("astral" planes) code points.
+// This also corresponds to the split between one-UTF-16 code unit
+// character and surrogate pairs, which gives us a natural
+// branch in the string parsing code.
+//
+// The an table is built which allows these chunks to overlap
```
Comment (natebosch) [inline tool/bin/gentable.dart]: Then a table?

File: tool/bin/gentable.dart [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+///
+/// Set while developing if you don't want to pass `-v` on the command line
+/// every time.
+/// Only affects this file when it's run directly.
+const defaultVerbose = false;
+
+/// Default location for table file.
+const tableFile = "lib/src/grapheme_clusters/table.dart";
+
+/// Best values found for current tables.
```
Comment (natebosch) [inline tool/bin/gentable.dart]: This doesn't look like a doc for a specific member - 2 slashes here?

File: tool/bin/gentest.dart [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    stdout.write(buffer);
+  } else {
+    if (verbose) {
+      stderr.writeln("Writing ${output.path}");
+    }
+    output.writeAsStringSync(buffer.toString());
+  }
+}
+
+bool _escape(int cp) => cp > 0xff || cp & 0x60 == 0 || cp == 0x7f;
```
Comment (natebosch) [inline tool/bin/gentest.dart:131]: [nit] how about `_needsEscape` or `_shouldEscape`?

File: tool/src/args.dart [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    } else if (arg == "-u" || arg == "--update") {
+      update = true;
+    } else if (arg == "-n" || arg == "--dryrun") {
+      dryrun = true;
+    } else if (arg == "-v" || arg == "--verbose") {
+      verbose = true;
+    } else if (allowOptimize && arg == "-o" ||
+        arg == "-i" ||
+        arg.startsWith("--opt")) {
+      /// Try to find a better size for the table.
```
Comment (natebosch) [inline tool/src/args.dart]: [nit] 2 slashes

File: tool/src/atsp.dart [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+/// This is an implementation of step of 3-opt, a simple algorithm to
+/// approximate an asymmetric traveling salesman problem (ATSP).
+/// It splits the cycle into three parts and then find the best recombination
+/// of the parts, each potentially reversed.
+bool opt3(Graph graph, List<int> cycle) {
+  // Perhaps optimize the weight computations by creating
+  // a single array of cummulative weights, so any range can be computed
+  // as a difference between two points in that array.
+  for (int i = 1; i < cycle.length; i++) {
+    int wA = cycle[i - 1];
```
Comment (natebosch) [inline tool/src/atsp.dart]: What is the `w` here short for?

I understand it is "weight" in `wAB`, but I would have expected a word like `vertexA` for this variable.

File: tool/src/atsp.dart [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+// where the distance between chunks is how little they overlap
+// (chunk length minus overlap size).
+
+/// An asymmetric weighted complete graph.
+///
+/// The vertices are identified by numbers 0 through [vertices] - 1.
+/// Edges are pairs of vertices.
+class Graph {
+  /// Number of vertices.
+  final int vertices;
```
Comment (natebosch) [inline tool/src/atsp.dart]: [optional] How about `vertexCount`?

File: tool/src/grapheme_category_loader.dart [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+Uint8List _parseCategories(List<String> files, {required bool verbose}) {
+  var result = Uint8List(0x110000);
+  result.fillRange(0, result.length, categoryOther);
+  int count = 0;
+  var categoryCount = <String, int>{};
+  var categoryMin = <String, int>{
+    for (var category in categoryByName.keys) category: 0x10FFFF
+  };
+  int min(int a, int b) => a < b ? a : b;
```
Comment (natebosch) [inline tool/src/grapheme_category_loader.dart:61]: Import from `dart:math`?

File: tool/src/shared.dart [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+Future<String> fetch(String location,
+    {File? targetFile, bool forceLoad = false}) async {
+  if (targetFile == null) {
+    String safeLocation = location.replaceAll(RegExp(r"[^\w]+"), "-");
+    targetFile = File(path(Directory.systemTemp.path, safeLocation));
+  }
+  if (!forceLoad && targetFile.existsSync()) {
+    return targetFile.readAsString();
+  }
+  Uri uri = Uri.parse(location);
```
Comment (natebosch) [inline tool/src/shared.dart]: We should consider turning on omit_local_variable_types and cleaning this up. I could do that in a followup.

**Review by natebosch — APPROVED** [review]
There was a lot here, and since this is tooling code I didn't review as fully as I would if it was user facing. This looks pretty good overall.

File: tool/src/atsp.dart (hunk 6 above)
Comment (lrhn) [reply tool/src/atsp.dart]: Changed to `nodeA`, `wAB` and `pBC` to distinguish single-step weights from path weights, and nodes from weights.

File: tool/src/grapheme_category_loader.dart (hunk 8 above)
Comment (lrhn) [reply tool/src/grapheme_category_loader.dart:61]: That one is wasteful because it works for `double` too.

File: tool/src/atsp.dart (hunk 7 above)
Comment (lrhn) [reply tool/src/atsp.dart]: The way the class is written, the integers *are* the vertices.

But done.

File: tool/src/shared.dart (hunk 9 above)
Comment (lrhn) [reply tool/src/shared.dart]: Done.

**Comment by lrhn** [comment]
Addressed comments, switched to more lints.

Didn't generate new files for now, since data files haven't changed.

**Review by natebosch — APPROVED** [review]

**Pull request merged**